File: backend/billing_app/invoices/services/calculator.py
"""
Invoice calculation service for computing totals, taxes, and levies.

This module provides precise decimal-based calculations for invoice totals,
including subtotals, tax/levy amounts, and grand totals. All calculations
use Python's Decimal type to avoid floating-point precision issues common
in financial calculations.
"""
from dataclasses import dataclass
from decimal import Decimal, ROUND_HALF_UP
from typing import Dict, Iterable

from django.conf import settings


@dataclass
class InvoiceTotals:
    """
    Data class holding calculated invoice totals.
    
    Attributes:
        subtotal: Sum of all line items (before taxes/levies)
        levies: Dictionary mapping levy name to calculated amount
        grand_total: Final total including subtotal and all levies
    """
    subtotal: Decimal
    levies: Dict[str, Decimal]
    grand_total: Decimal


def _to_decimal(value) -> Decimal:
    """
    Convert numeric value to Decimal with 2 decimal places.
    
    Safely converts various numeric types (int, float, str, Decimal)
    to Decimal with proper rounding. Uses ROUND_HALF_UP to round
    values like 0.125 to 0.13 (traditional rounding).
    
    Args:
        value: Numeric value to convert (int, float, str, or Decimal)
    
    Returns:
        Decimal: Value converted to Decimal with 2 decimal places
    """
    # If already Decimal, return as-is
    if isinstance(value, Decimal):
        return value
    # Convert to string first to avoid float precision issues, then to Decimal
    return Decimal(str(value or 0)).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)


def calculate_totals(items: Iterable[dict]) -> InvoiceTotals:
    """
    Calculate invoice subtotal, levies, and grand total from line items.
    
    Computes:
    1. Subtotal: Sum of (quantity × unit_price) for all items
    2. Levies: Tax amounts calculated as percentages of subtotal
    3. Grand Total: Subtotal plus all levies
    
    Tax/levy rates are read from settings.TAX_SETTINGS, which should be
    a dictionary mapping levy names to rates (e.g., {"VAT": 0.15} for 15%).
    
    Args:
        items: Iterable of item dictionaries, each with keys:
               - quantity: Item quantity (numeric)
               - unit_price: Price per unit (numeric)
    
    Returns:
        InvoiceTotals: Object containing subtotal, levies dict, and grand_total
    
    Example:
        >>> items = [{"quantity": 2, "unit_price": 10.50}]
        >>> totals = calculate_totals(items)
        >>> print(totals.subtotal)  # Decimal('21.00')
    """
    # Calculate subtotal: sum of (quantity × unit_price) for all items
    subtotal = sum(
        _to_decimal(item.get("quantity", 0)) * _to_decimal(item.get("unit_price", 0))
        for item in items or []
    ).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)

    # Calculate each levy/tax as percentage of subtotal
    levies: Dict[str, Decimal] = {}
    levy_total = Decimal("0.00")
    for levy_name, rate in settings.TAX_SETTINGS.items():
        # Calculate levy amount: subtotal × rate (e.g., subtotal × 0.15 for 15% VAT)
        levy_amount = (subtotal * Decimal(str(rate))).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
        levies[levy_name] = levy_amount
        levy_total += levy_amount

    # Calculate grand total: subtotal + all levies
    grand_total = subtotal + levy_total
    grand_total = grand_total.quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)

    return InvoiceTotals(subtotal=subtotal, levies=levies, grand_total=grand_total)

File: billdoc/lib/services/calculator.dart
import 'package:decimal/decimal.dart';
import '../models/tax_settings.dart';

/// Invoice totals result matching backend InvoiceTotals dataclass
class InvoiceTotals {
  final Decimal subtotal;
  final Map<String, Decimal> levies;
  final Decimal grandTotal;

  const InvoiceTotals({
    required this.subtotal,
    required this.levies,
    required this.grandTotal,
  });

  /// Convert to JSON for API compatibility
  Map<String, dynamic> toJson() {
    return {
      'subtotal': subtotal.toDouble(),
      'levies': levies.map((key, value) => MapEntry(key, value.toDouble())),
      'grand_total': grandTotal.toDouble(),
    };
  }
}

/// Calculator service matching backend calculator.py logic
class Calculator {
  /// Helper to round Decimal to 2 decimal places (ROUND_HALF_UP)
  static Decimal _roundToTwoDecimals(Decimal value) {
    // Multiply by 100, round, then divide by 100
    // Convert from Rational back to Decimal
    final shifted = value * Decimal.fromInt(100);
    final rounded = shifted.round();
    final result = rounded / Decimal.fromInt(100);
    // Convert Rational to Decimal by going through string
    return Decimal.parse(result.toDouble().toStringAsFixed(2));
  }

  /// Convert value to Decimal with proper rounding (ROUND_HALF_UP)
  /// Matches backend _to_decimal function
  static Decimal toDecimal(dynamic value) {
    if (value is Decimal) {
      return _roundToTwoDecimals(value);
    }
    final numValue = value is num ? value : (double.tryParse(value.toString()) ?? 0.0);
    final decimal = Decimal.parse(numValue.toString());
    return _roundToTwoDecimals(decimal);
  }

  /// Calculate invoice totals matching backend calculate_totals function
  /// Uses Decimal arithmetic with ROUND_HALF_UP (default in Decimal package)
  static InvoiceTotals calculateTotals(
    List<Map<String, dynamic>> items,
    TaxSettings taxSettings,
  ) {
    // Calculate subtotal: sum of (quantity * unit_price) for all items
    Decimal subtotal = Decimal.zero;
    for (final item in items) {
      final quantity = toDecimal(item['quantity'] ?? 0);
      final unitPrice = toDecimal(item['unit_price'] ?? 0);
      final itemTotal = quantity * unitPrice;
      subtotal += _roundToTwoDecimals(itemTotal);
    }
    subtotal = _roundToTwoDecimals(subtotal);

    // Calculate levies (taxes) based on tax settings
    final levies = <String, Decimal>{};
    Decimal levyTotal = Decimal.zero;

    for (final entry in taxSettings.asMap.entries) {
      final levyName = entry.key;
      final rate = toDecimal(entry.value);
      final levyAmount = subtotal * rate;
      final roundedLevy = _roundToTwoDecimals(levyAmount);
      levies[levyName] = roundedLevy;
      levyTotal += roundedLevy;
    }

    // Calculate grand total
    final grandTotal = subtotal + levyTotal;
    final roundedGrandTotal = _roundToTwoDecimals(grandTotal);

    return InvoiceTotals(
      subtotal: subtotal,
      levies: levies,
      grandTotal: roundedGrandTotal,
    );
  }

  /// Calculate total for a single item
  static double calculateItemTotal(double quantity, double unitPrice) {
    final result = toDecimal(quantity) * toDecimal(unitPrice);
    return result.toDouble();
  }
}